src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -0,0 +1,226 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Holds on a <see cref="DocumentId"/> to <see cref="TextDocumentState"/> map and an ordering.
+    /// </summary>
+    internal readonly struct TextDocumentStates<TState>
+        where TState : TextDocumentState
```

Comment:
should this implement ienumerable?  So you don't need to do DocumentStates.States everywhere you want to iterate them (or would that be a perf issue)

Reply:
It could be enumerable. However, some code enumerates `Ids` and some code enumerates `States`. These enumerations have different ordering (currently) and perf characteristic. I think it's better to be explicit. If we switch to different underlying data structures and we can just use a single ordering that's fast then I think it would make sense to make this type directly enumerable.